File: LeetCode/0085/golang/lc0085.go
package lc0085

import "fmt"

type Stack []int

func (s *Stack) Push(val int) {
	*s = append(*s, val)
}

func (s *Stack) Pop() (int, error) {
	sz := len(*s)
	if sz == 0 {
		return 0, fmt.Errorf("Pop() from an empty Stack.")
	}
	val := (*s)[sz-1]
	*s = (*s)[:sz-1]
	return val, nil
}

func (s *Stack) Top() int {
	sz := len(*s)
	if sz == 0 {
		return -42 // only ever used in conjunction with IsEmpty(). Will never hit.
	}
	return (*s)[sz-1]
}

func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}

// Runtime complexity: O(H*W)
// Auxiliary space: O(W)
// Subjective level: hard
// Solved on: 2026-01-11
func maximalRectangle(matrix [][]byte) int {
	//H := len(matrix)
	W := len(matrix[0])
	histogram := make([]int, W)

	getLargestRectangleArea := func() int {
		area := 0
		stack := Stack{}
		for h := 0; h <= len(histogram); h++ {
			for !stack.IsEmpty() && (h == len(histogram) || histogram[stack.Top()] > histogram[h]) {
				height := histogram[stack.Top()]
				stack.Pop()
				width := 0
				if stack.IsEmpty() {
					width = h
				} else {
					width = h - stack.Top() - 1
				}
				area = max(area, height*width)
			}
			stack.Push(h)
		}
		return area
	}

	ans := 0
	for _, row := range matrix {
		for w := range W {
			if row[w] == '0' {
				histogram[w] = 0
			} else {
				histogram[w] = histogram[w] + 1
			}
		}
		ans = max(ans, getLargestRectangleArea())
	}
	return ans
}

File: LeetCode/0085/cpp/lc0085.cpp
#include <cassert>
#include <stack>
#include <vector>

class Solution {
public:
  int maximalRectangle(std::vector<std::vector<char>>& matrix) {
    const int HEIGHT = matrix.size();
    const int WIDTH = matrix[0].size();
    std::vector<int> histogram(WIDTH);
    int ans = 0;

    for (const std::vector<char>& row: matrix) {
      for (int w = 0; w < WIDTH; ++w) {
        histogram[w] = (row[w] == '0') ? 0 : histogram[w] + 1;
      }
      ans = std::max(ans, getLargestRectangleArea(histogram));
    }

    return ans;
  }

private:
  int getLargestRectangleArea(const std::vector<int>& hist) {
    int ans = 0;
    std::stack<int> stack;
    for (int i = 0; i <= hist.size(); ++i) {
      while (
        !stack.empty() &&
        (i == hist.size() || hist[stack.top()] > hist[i])
      ) {
        const int height = hist[stack.top()];
        stack.pop();
        const int width = (stack.empty()) ? i : i - stack.top() - 1;
        ans = std::max(ans, height * width);
      }
      stack.push(i);
    }
    return ans;
  }
};

int main() {
  Solution s;
  {
    std::vector<std::vector<char>> matrix = {
      {'1', '0', '1', '0', '0'},
      {'1', '0', '1', '1', '1'},
      {'1', '1', '1', '1', '1'},
      {'1', '0', '0', '1', '0'},
    };
    assert(s.maximalRectangle(matrix) == 6);
  }
  {
    std::vector<std::vector<char>> matrix = {
      {'0'},
    };
    assert(s.maximalRectangle(matrix) == 0);
  }
  {
    std::vector<std::vector<char>> matrix = {
      {'1'},
    };
    assert(s.maximalRectangle(matrix) == 1);
  }
}
